File: iOS/Tyche/Tyche/UI/Profile/AvatarCropView.swift
import SwiftUI
import UI

let AVATAR_MORPH_ID = "profileAvatar"

/// Pan + pinch-zoom crop over a fixed circular window on a near-black surface.
/// The image always covers the window; zoom is bounded by `AvatarCrop`. A dual-scale
/// preview strip below the window shows the crop at Profile size and at row size.
struct AvatarCropView: View {
    let image: UIImage
    let username: String
    let morphNamespace: Namespace.ID
    let onConfirm: (CGRect) -> Void
    let onCancel: () -> Void

    @State private var steadyScale: CGFloat?
    @State private var steadyOffset: CGSize = .zero
    @GestureState private var pinch: CGFloat = 1
    @GestureState private var drag: CGSize = .zero
    @State private var isAtBoundary = false

    @Environment(\.boxSpacing) private var boxSpacing

    private let boundaryHaptic = UIImpactFeedbackGenerator(style: .light)

    private var imagePixelSize: CGSize {
        CGSize(width: image.size.width * image.scale, height: image.size.height * image.scale)
    }

    var body: some View {
        GeometryReader { proxy in
            let windowSide = min(proxy.size.width - boxSpacing.huge, proxy.size.height * 0.5)
            let crop = AvatarCrop(imageSize: imagePixelSize, windowSide: windowSide)
            let baseScale = steadyScale ?? crop.minScale
            let liveScale = crop.clampScale(baseScale * pinch)
            let liveOffset = liveOffset(crop: crop, baseScale: baseScale, liveScale: liveScale)

            VStack(spacing: 0) {
                cropWindow(crop: crop, scale: liveScale, offset: liveOffset)
                    .frame(maxHeight: .infinity)

                previewStrip(scale: liveScale, offset: liveOffset, windowSide: windowSide)
                    .padding(.vertical, boxSpacing.extraLarge)

                actionBar(crop: crop, scale: liveScale, offset: liveOffset)
            }
            .background(surfaceColor.ignoresSafeArea())
            .contentShape(Rectangle())
            .gesture(cropGesture(crop: crop, baseScale: baseScale))
        }
    }

    private func liveOffset(crop: AvatarCrop, baseScale: CGFloat, liveScale: CGFloat) -> CGSize {
        let zoomedOffset = crop.offset(
            zoomingTo: liveScale,
            from: baseScale,
            offset: steadyOffset,
            anchor: .zero
        )

        return crop.clampOffset(
            CGSize(width: zoomedOffset.width + drag.width, height: zoomedOffset.height + drag.height),
            scale: liveScale
        )
    }

    // Built as overlays on Color.clear so the zoomed image can never inflate the layout —
    // a frame carrying the child's own size would blow the screen apart at high zoom.
    private func cropWindow(crop: AvatarCrop, scale: CGFloat, offset: CGSize) -> some View {
        Color.clear
            .overlay(transformedImage(scale: scale, offset: offset))
            .overlay(
                scrimColor
                    .mask {
                        Rectangle()
                            .overlay(
                                Circle()
                                    .frame(width: crop.windowSide, height: crop.windowSide)
                                    .blendMode(.destinationOut)
                            )
                            .compositingGroup()
                    }
            )
            .overlay(
                transformedImage(scale: scale, offset: offset)
                    .frame(width: crop.windowSide, height: crop.windowSide)
                    .clipShape(Circle())
                    .matchedGeometryEffect(id: AVATAR_MORPH_ID, in: morphNamespace)
            )
            .overlay(
                Circle()
                    .strokeBorder(hairlineColor, lineWidth: 1)
                    .frame(width: crop.windowSide, height: crop.windowSide)
            )
            .clipped()
    }

    private func transformedImage(scale: CGFloat, offset: CGSize) -> some View {
        Image(uiImage: image)
            .resizable()
            .frame(width: imagePixelSize.width * scale, height: imagePixelSize.height * scale)
            .offset(offset)
    }

    private func previewStrip(scale: CGFloat, offset: CGSize, windowSide: CGFloat) -> some View {
        HStack(spacing: boxSpacing.extraLarge) {
            cropPreview(scale: scale, offset: offset, windowSide: windowSide, diameter: profilePreviewSize)

            HStack(spacing: boxSpacing.medium) {
                cropPreview(scale: scale, offset: offset, windowSide: windowSide, diameter: rowPreviewSize)

                Text(username)
                    .font(.subheadline)
                    .foregroundStyle(previewLabelColor)
                    .lineLimit(1)
            }
        }
        .padding(.horizontal, boxSpacing.large)
    }

    private func cropPreview(scale: CGFloat, offset: CGSize, windowSide: CGFloat, diameter: CGFloat) -> some View {
        let previewFactor = diameter / windowSide

        return Image(uiImage: image)
            .resizable()
            .frame(
                width: imagePixelSize.width * scale * previewFactor,
                height: imagePixelSize.height * scale * previewFactor
            )
            .offset(CGSize(width: offset.width * previewFactor, height: offset.height * previewFactor))
            .frame(width: diameter, height: diameter)
            .clipShape(Circle())
    }

    private func actionBar(crop: AvatarCrop, scale: CGFloat, offset: CGSize) -> some View {
        HStack {
            Button(action: onCancel) {
                Text(sharedResource: .cancelAction)
                    .foregroundStyle(.white)
            }

            Spacer()

            Button {
                onConfirm(crop.cropRect(scale: scale, offset: offset))
            } label: {
                Text(.usePhotoAction)
                    .fontWeight(.semibold)
            }
            .buttonStyle(.borderedProminent)
        }
        .padding(.horizontal, boxSpacing.large)
        .padding(.bottom, boxSpacing.large)
    }

    private func cropGesture(crop: AvatarCrop, baseScale: CGFloat) -> some Gesture {
        let dragGesture = DragGesture()
            .updating($drag) { value, state, _ in
                state = value.translation
            }
            .onChanged { value in
                let liveScale = crop.clampScale(baseScale * pinch)
                let zoomedOffset = crop.offset(
                    zoomingTo: liveScale,
                    from: baseScale,
                    offset: steadyOffset,
                    anchor: .zero
                )
                let rawOffset = CGSize(
                    width: zoomedOffset.width + value.translation.width,
                    height: zoomedOffset.height + value.translation.height
                )
                tickOnClamp(raw: rawOffset, clamped: crop.clampOffset(rawOffset, scale: liveScale))
            }

        let pinchGesture = MagnificationGesture()
            .updating($pinch) { value, state, _ in
                state = value
            }
            .onChanged { value in
                let rawScale = baseScale * value
                tickOnClamp(raw: rawScale, clamped: crop.clampScale(rawScale))
            }

        // One commit for both gestures, mirroring the live-display math exactly —
        // per-gesture commits would re-apply the zoom ratio to an already-committed pan.
        return dragGesture
            .simultaneously(with: pinchGesture)
            .onEnded { value in
                let translation = value.first?.translation ?? .zero
                let newScale = crop.clampScale(baseScale * (value.second ?? 1))
                let zoomedOffset = crop.offset(
                    zoomingTo: newScale,
                    from: baseScale,
                    offset: steadyOffset,
                    anchor: .zero
                )

                steadyOffset = crop.clampOffset(
                    CGSize(
                        width: zoomedOffset.width + translation.width,
                        height: zoomedOffset.height + translation.height
                    ),
                    scale: newScale
                )
                steadyScale = newScale
                isAtBoundary = false
            }
    }

    private func tickOnClamp(raw: CGSize, clamped: CGSize) {
        tick(clamping: raw != clamped)
    }

    private func tickOnClamp(raw: CGFloat, clamped: CGFloat) {
        tick(clamping: raw != clamped)
    }

    private func tick(clamping: Bool) {
        if clamping && !isAtBoundary {
            boundaryHaptic.impactOccurred()
        }
        isAtBoundary = clamping
    }
}

private let profilePreviewSize: CGFloat = 96
private let rowPreviewSize: CGFloat = 32

private let surfaceColor = Color(red: 0.04, green: 0.04, blue: 0.04)
private let scrimColor = Color.black.opacity(0.62)
private let hairlineColor = Color.white.opacity(0.24)
private let previewLabelColor = Color.white.opacity(0.85)

#Preview {
    struct CropPreviewHost: View {
        @Namespace private var namespace

        var body: some View {
            AvatarCropView(
                image: UIImage(systemName: "photo")!,
                username: "ElGoleador",
                morphNamespace: namespace,
                onConfirm: { _ in },
                onCancel: {}
            )
        }
    }

    return CropPreviewHost()
}

File: Android/app/src/main/java/com/felipearpa/tyche/profile/AvatarCropView.kt
package com.felipearpa.tyche.profile

import android.graphics.Bitmap
import android.view.HapticFeedbackConstants
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.gestures.detectTransformGestures
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.requiredSize
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Button
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.clipToBounds
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.CompositingStrategy
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.platform.LocalView
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.IntOffset
import androidx.compose.ui.unit.dp
import com.felipearpa.tyche.R
import com.felipearpa.tyche.ui.theme.LocalBoxSpacing
import androidx.compose.foundation.Canvas as ComposeCanvas
import com.felipearpa.tyche.ui.R as SharedR
import kotlin.math.min
import kotlin.math.roundToInt

/**
 * Pan + pinch-zoom crop over a fixed circular window on a near-black surface.
 * The image always covers the window; zoom is bounded by [AvatarCrop] and anchored at the
 * pinch centroid. A dual-scale preview strip below the window shows the crop at Profile
 * size and at row size.
 */
@Composable
fun AvatarCropView(
    bitmap: Bitmap,
    username: String,
    onConfirm: (androidx.compose.ui.geometry.Rect) -> Unit,
    onCancel: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val imageSize = Size(bitmap.width.toFloat(), bitmap.height.toFloat())
    val view = LocalView.current

    var steadyScale by remember(bitmap) { mutableFloatStateOf(0f) }
    var steadyOffset by remember(bitmap) { mutableStateOf(Offset.Zero) }
    var isAtBoundary by remember { mutableStateOf(false) }

    Column(
        modifier = modifier
            .fillMaxSize()
            .background(SURFACE_COLOR)
            // The screen is edge-to-edge; without this the action bar lands inside the system
            // gesture-navigation zone, which swallows taps.
            .navigationBarsPadding(),
    ) {
        BoxWithConstraints(
            modifier = Modifier
                .fillMaxWidth()
                .weight(1f),
        ) {
            val density = LocalDensity.current
            val margin = with(density) { LocalBoxSpacing.current.huge.toPx() }
            // The height cap keeps the circle inside the gesture area above the preview strip
            // and action bar, matching the iOS sizing rule.
            val windowSide = min(
                constraints.maxWidth.toFloat() - margin,
                constraints.maxHeight.toFloat() * 0.5f,
            )
            val crop = AvatarCrop(imageSize = imageSize, windowSide = windowSide)
            val scale = if (steadyScale == 0f) crop.minScale else crop.clampScale(steadyScale)
            val offset = crop.clampOffset(steadyOffset, scale)

            Column(modifier = Modifier.fillMaxSize()) {
                Box(
                    modifier = Modifier
                        .fillMaxWidth()
                        .weight(1f)
                        .clipToBounds()
                        .pointerInput(crop) {
                            detectTransformGestures { centroid, pan, zoom, _ ->
                                val currentScale =
                                    if (steadyScale == 0f) crop.minScale else steadyScale
                                val anchor = centroid - Offset(
                                    size.width / 2f,
                                    size.height / 2f,
                                )

                                val wantedScale = currentScale * zoom
                                val newScale = crop.clampScale(wantedScale)

                                val zoomedOffset = crop.offset(
                                    zoomingTo = newScale,
                                    from = currentScale,
                                    offset = steadyOffset,
                                    anchor = anchor,
                                )
                                val wantedOffset = zoomedOffset + pan
                                val newOffset = crop.clampOffset(wantedOffset, newScale)

                                val clamping =
                                    wantedScale != newScale || wantedOffset != newOffset
                                if (clamping && !isAtBoundary) {
                                    view.performHapticFeedback(HapticFeedbackConstants.CLOCK_TICK)
                                }
                                isAtBoundary = clamping

                                steadyScale = newScale
                                steadyOffset = newOffset
                            }
                        },
                ) {
                    TransformedImage(
                        bitmap = bitmap,
                        imageSize = imageSize,
                        scale = scale,
                        offset = offset,
                        modifier = Modifier.align(Alignment.Center),
                    )

                    ComposeCanvas(
                        modifier = Modifier
                            .fillMaxSize()
                            .graphicsLayer { compositingStrategy = CompositingStrategy.Offscreen },
                    ) {
                        drawRect(color = SCRIM_COLOR)
                        drawCircle(
                            color = Color.Transparent,
                            radius = windowSide / 2f,
                            blendMode = BlendMode.Clear,
                        )
                        drawCircle(
                            color = HAIRLINE_COLOR,
                            radius = windowSide / 2f,
                            style = Stroke(width = 1.dp.toPx()),
                        )
                    }

                }

                PreviewStrip(
                    bitmap = bitmap,
                    imageSize = imageSize,
                    scale = scale,
                    offset = offset,
                    windowSide = windowSide,
                    username = username,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(vertical = LocalBoxSpacing.current.extraLarge),
                )

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(
                            horizontal = LocalBoxSpacing.current.large,
                            vertical = LocalBoxSpacing.current.large,
                        ),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    TextButton(onClick = onCancel) {
                        Text(
                            text = stringResource(id = SharedR.string.cancel_action),
                            color = Color.White,
                        )
                    }

                    Button(onClick = { onConfirm(crop.cropRect(scale = scale, offset = offset)) }) {
                        Text(text = stringResource(id = R.string.use_photo_action))
                    }
                }
            }
        }
    }
}

@Composable
private fun TransformedImage(
    bitmap: Bitmap,
    imageSize: Size,
    scale: Float,
    offset: Offset,
    modifier: Modifier = Modifier,
) {
    val density = LocalDensity.current

    Image(
        bitmap = bitmap.asImageBitmap(),
        contentDescription = null,
        contentScale = ContentScale.FillBounds,
        modifier = modifier
            .offset { IntOffset(offset.x.roundToInt(), offset.y.roundToInt()) }
            .requiredSize(
                width = with(density) { (imageSize.width * scale).toDp() },
                height = with(density) { (imageSize.height * scale).toDp() },
            ),
    )
}

@Composable
private fun PreviewStrip(
    bitmap: Bitmap,
    imageSize: Size,
    scale: Float,
    offset: Offset,
    windowSide: Float,
    username: String,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(
            LocalBoxSpacing.current.extraLarge,
            Alignment.CenterHorizontally,
        ),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CropPreview(
            bitmap = bitmap,
            imageSize = imageSize,
            scale = scale,
            offset = offset,
            windowSide = windowSide,
            diameter = PROFILE_PREVIEW_SIZE.dp,
        )

        Row(
            horizontalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.medium),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            CropPreview(
                bitmap = bitmap,
                imageSize = imageSize,
                scale = scale,
                offset = offset,
                windowSide = windowSide,
                diameter = ROW_PREVIEW_SIZE.dp,
            )

            Text(
                text = username,
                style = MaterialTheme.typography.bodyMedium,
                color = PREVIEW_LABEL_COLOR,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }
    }
}

@Composable
private fun CropPreview(
    bitmap: Bitmap,
    imageSize: Size,
    scale: Float,
    offset: Offset,
    windowSide: Float,
    diameter: androidx.compose.ui.unit.Dp,
    modifier: Modifier = Modifier,
) {
    val density = LocalDensity.current
    val previewFactor = with(density) { diameter.toPx() } / windowSide

    Box(
        modifier = modifier
            .size(diameter)
            .clip(CircleShape),
    ) {
        TransformedImage(
            bitmap = bitmap,
            imageSize = imageSize,
            scale = scale * previewFactor,
            offset = offset * previewFactor,
            modifier = Modifier.align(Alignment.Center),
        )
    }
}

private const val PROFILE_PREVIEW_SIZE = 96
private const val ROW_PREVIEW_SIZE = 32

private val SURFACE_COLOR = Color(0xFF0A0A0A)
private val SCRIM_COLOR = Color.Black.copy(alpha = 0.62f)
private val HAIRLINE_COLOR = Color.White.copy(alpha = 0.24f)
private val PREVIEW_LABEL_COLOR = Color.White.copy(alpha = 0.85f)
